### ISSUE TITLE
EqualsTopologically works wrong for geometry collections

### ISSUE THREAD

**Issue by grisha0088** [NetTopologySuite/NetTopologySuite#213]
EqualsTopologically works well only if passed IGeomentry param is homogenous. If it is geometry collection it should compare all geometries in collection.

**Comment by roji** [comment]
Got bitten by this as the EF Core tracking mechanism calls EqualsTopologically to check whether two instances of a geometry are equal (and therefore no changes need to be written to the database).

**Comment by roji** [comment]
Is the correct solution here to simply compare geometries one-by-one? Since from a topological point of view, a geometry with 2 elements may be equal to a geometry with 3 elements since they cover exactly the same space. Seems like a union of the collection items should be done and the comparison done on that?

**Comment by airbreather** [comment]
> Seems like a union of the collection items should be done and the comparison done on that?

I don't think so.  Even if the elements were just `Polygon`s, consider the following case:
![image](https://user-images.githubusercontent.com/3792815/68943542-78f54a80-0779-11ea-802e-52dc9443cfdd.png)

Two triangles share a segment.  The boundary of the `GeometryCollection` with both polygons is different from the boundary of their union, so not all predicates will return the same result.  For example, in MSSQL:

```sql
DECLARE @coll geometry = geometry::STGeomFromText('GEOMETRYCOLLECTION(POLYGON ((0 0, 0 1, 1 0, 0 0)), POLYGON ((0 1, 1 1, 1 0, 0 1)))', 0)
DECLARE @sharedSegment geometry = geometry::STGeomFromText('LINESTRING (0 1, 1 0)', 0)

SELECT
	OriginalCollectionContains = @coll.STContains(@sharedSegment),
	UnionContains = geometry::UnionAggregate(@coll).STContains(@sharedSegment)
```

This selects:

|OriginalCollectionContains|UnionContains|
|-|-|
|0|1|

**Comment by FObermaier** [comment]
We could do this: 
* Extract the single instance geometries (`Point`, `LineString`, `Polygon`) from the `GeometryCollection`.
* Sort them by type and build `Multi`-geometries. 
* Call `EqualsTopologically` on each of these.

**Comment by archanox** [comment]
Is this road mapped at all? Or is there an ETA for this? Or an appropriate work around?
Unfortunately this is a major blocker.

**Comment by archanox** [comment]
@airbreather is there an upstream issue or branch that we can follow the progress on?

**Comment by airbreather** [comment]
> @airbreather is there an upstream issue or branch that we can follow the progress on?

@archanox opened locationtech/jts#539

**Comment by dr-jts** [comment]
> Got bitten by this as the EF Core tracking mechanism calls EqualsTopologically to check whether two instances of a geometry are equal (and therefore no changes need to be written to the database).

Why do you want to use topologically equality as the basis for determining whether database update is required?  It's a fairly weak condition, and seems like not at all what a user would expect for update semantics.

**Comment by roji** [comment]
FWIW I'm not convinced that it's the right thing - in https://github.com/dotnet/efcore/issues/18921 we actually switched from topological to exact, though that was also partially driven by the lack of topological equality for collections.

Am not spatial expert at all, but it seems that non-topological differences (like ordering in the collections) would be largely irrelevant, as operators would return the same results for topologically equal geometries. What's your case for thinking that exact is more appropriate for the general case?

**Comment by dr-jts** [comment]
Actually I agree that *exact* equality is too restrictive. A better equality condition might be *normalized equality*, which is insensitive to geometry ordering.  (Actually full normalized equality is also insensitive to linestring orientation, which is probably not what you want, since line orientation can often carry information).

Perhaps the best approach would be an equality check which is insensitive to the ordering of geometries in a collection, but does check for exact equality against "matching" pairs of geometry.

**Comment by roji** [comment]
That makes sense to me, even if I don't fully understand the details (again am a total spatial outsider).

**Comment by archanox** [comment]
Is there a rough ETA when this issue will be resolved?

**Comment by airbreather** [comment]
Per https://github.com/locationtech/jts/issues/539#issuecomment-2574190919 - see if RelateNG does something consistent and correct enough to close this (and the upstream request)
